common/beerocks/bcl/source/beerocks_event_loop_impl.cpp

```diff
@@ -0,0 +1,215 @@
+
+// Maximal number of events to process in a single epoll_wait call
+static constexpr int MAX_POLL_EVENTS = 17;
+
+//////////////////////////////////////////////////////////////////////////////
+/////////////////////////////// Implementation ///////////////////////////////
+//////////////////////////////////////////////////////////////////////////////
+
+EventLoopImpl::EventLoopImpl(std::chrono::milliseconds timeout) : m_timeout(timeout)
+{
+    m_epoll_fd = epoll_create1(0);
+    LOG_IF(m_epoll_fd == -1, FATAL) << "Failed creating epoll: " << strerror(errno);
+}
+
+EventLoopImpl::~EventLoopImpl()
+{
+    // Delete all the file descriptors in the poll
+    LOG(DEBUG) << "Removing " << m_fd_to_event_handlers.size() << " FDs from the event loop";
+
+    while (!m_fd_to_event_handlers.empty()) {
+        int fd = m_fd_to_event_handlers.begin()->first;
+        EventLoopImpl::remove_handlers(fd);
+    }
+
+    // Close the poll fd
+    LOG_IF(close(m_epoll_fd) == -1, ERROR)
+        << "Failed closing epoll file descriptor: " << strerror(errno);
+}
+
+bool EventLoopImpl::register_handlers(int fd, const EventLoop::EventHandlers &handlers)
+{
+    if (-1 == fd) {
+        LOG(ERROR) << "Invalid file descriptor!";
+        return false;
+    }
+
+    // Make sure that the file descriptor is not already part of the poll
+    if (m_fd_to_event_handlers.find(fd) != m_fd_to_event_handlers.end()) {
+        LOG(WARNING) << "Requested to add FD (" << fd << ") to the poll, but it's already there...";
+
+        return false;
+    }
+
+    // Helper lambda function for adding a fd to the poll, and register for the following events:
+    // EPOLLIN: The associated fd is available for read operations.
+    // EPOLLOUT: The associated fd is available for write operations.
+    // EPOLLRDHUP: Socket peer closed connection, or shut down writing half of connection.
+    // EPOLLERR: Error condition happened on the associated fd.
+    // EPOLLHUP: Hang up happened on the associated fd.
+    auto add_fd_to_epoll = [&](int fd) -> bool {
+        epoll_event event = {};
+        event.data.fd     = fd;
+        event.events      = EPOLLRDHUP | EPOLLERR | EPOLLHUP;
+
+        // If read handler was set, also listen for POLL-IN events
+        if (handlers.on_read) {
+            event.events |= EPOLLIN;
+        }
+
+        // If write handler was set, also listen for POLL-OUT events
+        if (handlers.on_write) {
+            event.events |= EPOLLOUT;
+        }
+
+        if (epoll_ctl(m_epoll_fd, EPOLL_CTL_ADD, fd, &event) == -1) {
+            LOG(ERROR) << "Failed adding FD (" << fd << ") to the poll: " << strerror(errno);
+            return false;
+        }
+
+        // Map the file descriptor to the event handlers structure
+        m_fd_to_event_handlers[fd] = handlers;
+
+        return true;
+    };
+
+    // Add the file descriptor to the poll
+    if (!add_fd_to_epoll(fd)) {
+        return false;
+    }
```

Comment:
I like this part of code :)

Reply:
You should congratulate @vitalybu for it.
IMHO using a lambda here is not necessary and adds an extra level of indentation.

Reply:
But it groups related code together.

Reply:
Before removing the the timeout related source code, this lambda was called from two different places (that's why it was created in the first place)...